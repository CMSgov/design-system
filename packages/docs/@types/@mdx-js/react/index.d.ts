declare module '@mdx-js/react' {
  const MDXProvider: React.ComponentType;
  export { MDXProvider };
}
