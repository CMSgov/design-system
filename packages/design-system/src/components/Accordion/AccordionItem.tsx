import React from 'react';
import classNames from 'classnames';
import uniqueId from 'lodash.uniqueid';

export interface AccordionItemProps {
  /**
   * Class to be applied to the header `<button>` of an accordion item.
   */
  buttonClassName?: string;

  children?: React.ReactNode;
  /**
   * Class to be applied to the content `<div>` tag of an accordion item.
   */
  contentClassName?: string;
  /**
   * Boolean to expand the accordion.
   */
  defaultOpen?: boolean;
  /**
   * Text for the accordion item heading.
   */
  heading: React.ReactNode | string;
  /**
   *  Heading type to override default `<h2>`.
   */
  headingLevel?: '1' | '2' | '3' | '4' | '5' | '6';
  /**
   *  If not provided, a unique id will be automatically generated and used.
   */
  id?: string;
  /**
   * Sets the accordion panel's open state. Use this in combination with `onChange`
   * for a controlled accordion; otherwise, set `defaultOpen`.
   */
  isControlledOpen?: boolean;
  /**
   * A callback function that's invoked when a controlled accordion panel is selected or deselected.
   */
  onChange?: () => void;
}
export interface AccordionItemState {
  isOpen?: boolean;
}
export class AccordionItem extends React.Component<AccordionItemProps, AccordionItemState> {
  static defaultProps = {
    headingLevel: '2',
  };

  /* eslint-disable react/sort-comp */
  buttonId: string;
  contentId: string;
  isControlled: boolean;
  /* eslint-enable react/sort-comp */

  constructor(props: AccordionItemProps) {
    super(props);

    this.isControlled = !!props.onChange;
    this.state = this.isControlled ? {} : { isOpen: !!props.defaultOpen };
    this.handleClick = this.handleClick.bind(this);
    this.contentId = props.id || uniqueId('accordionItem_');
    this.buttonId = `${this.contentId}-button`;
  }

  // Set the state for opening and closing an accordion item
  handleClick(): void {
    if (this.isControlled) {
      this.props.onChange();
    } else {
      this.setState({ isOpen: !this.state.isOpen });
    }
  }

  render(): React.ReactNode {
    const {
      buttonClassName,
      children,
      contentClassName,
      heading,
      headingLevel = '2',
      isControlledOpen,
    } = this.props;
    const contentClasses = classNames('ds-c-accordion__content', contentClassName);
    const buttonClasses = classNames('ds-c-accordion__button', buttonClassName);
    const HeadingTag = `h${headingLevel}` as 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

    if (heading) {
      return (
        <>
          <HeadingTag className="ds-c-accordion__heading">
            <button
              className={buttonClasses}
              aria-expanded={this.isControlled ? isControlledOpen : this.state.isOpen}
              aria-controls={this.contentId}
              id={this.buttonId}
              onClick={this.handleClick}
            >
              {heading}
            </button>
          </HeadingTag>
          <div
            className={contentClasses}
            aria-labelledby={this.buttonId}
            id={this.contentId}
            hidden={this.isControlled ? !isControlledOpen : !this.state.isOpen}
          >
            {children}
          </div>
        </>
      );
    }
  }
}

export default AccordionItem;
