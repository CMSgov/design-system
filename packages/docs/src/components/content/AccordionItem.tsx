import { AccordionItem, MinusCircleIcon, PlusCircleIcon } from '@cmsgov/design-system';

AccordionItem.defaultProps = {
  ...AccordionItem.defaultProps,
  closeIconComponent: MinusCircleIcon,
  openIconComponent: PlusCircleIcon,
};

export default AccordionItem;
