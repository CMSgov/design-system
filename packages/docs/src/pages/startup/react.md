---
title: Using React components
weight: 3
---

The examples below assume you've installed the design system using NPM and have already setup your build system.

<h4 class="ds-h4 ds-u-font-size--base">Default imports</h4>

Individual components can be imported from their individual export file.

```jsx
import Button from '@cmsgov/design-system-core/dist/components/Button/Button';
```

<h4 class="ds-h4 ds-u-font-size--base">Named imports</h4>

Components can also be imported using the shorter syntax below.

<div class="ds-c-alert ds-c-alert--warn ds-u-margin-bottom--2 ds-u-font-size--small">
  <div class="ds-c-alert__body">
    <h3 class="ds-c-alert__heading ds-u-font-size--base">Performance note</h3>
    <p class="ds-c-alert__text">
      This approach may result in a much larger file than you intend. Depending on what module bundler you use, all of the design system's React components may be included in the bundled file even if you didn't specifically import them. This can be avoided by enabling features like [tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/).
    </p>
  </div>
</div>

```jsx
import { Button } from '@cmsgov/design-system-core';
```

[View an example](https://github.com/CMSgov/design-system/blob/master/examples/react-app/src/scripts/index.js)

<h2 id="examples" class="ds-h2 ds-u-color--primary-darker">Examples</h2>

Additional examples of the design system in use can be viewed on GitHub. These projects demonstrate the various ways you can incorporate the design system into your development process and various use cases.

<a href="https://github.com/CMSgov/design-system/tree/master/examples/" class="ds-c-button">Browse example projects</a>

<h2 id="need-help" class="ds-h2 ds-u-color--primary-darker">Need help or ran into an issue?</h2>

If you're having trouble installing or setting up the design system, or if you think you've found a bug, feel free to [open an issue on GitHub](https://github.com/CMSgov/design-system/issues).
