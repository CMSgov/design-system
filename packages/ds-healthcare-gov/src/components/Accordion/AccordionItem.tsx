import { AccordionItem, MinusCircleIcon, PlusCircleIcon } from '@cmsgov/design-system';

AccordionItem.defaultProps = {
  ...AccordionItem.defaultProps,
  closeIconComponent: MinusCircleIcon,
  openIconComponent: PlusCircleIcon,
};

// We need this for the healthcare story to work, and we also need to declare the side
// effects for the src path.
export { AccordionItem };
