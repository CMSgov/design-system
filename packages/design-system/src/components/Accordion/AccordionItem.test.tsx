import React from 'react';
import AccordionItem from './AccordionItem';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const defaultChildren = '<p>Content for accordion item</p>';
const defaultProps = {
  heading: 'Heading for accordion item',
};

function renderAccordionItem(customProps = {}) {
  const props = { ...defaultProps, ...customProps };

  return render(<AccordionItem {...props}>{defaultChildren}</AccordionItem>);
}

describe('AccordionItem', function () {
  it('renders an accordion item', () => {
    const { asFragment } = renderAccordionItem();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders an accordion heading', () => {
    renderAccordionItem();

    const headerEls = screen.getAllByRole('heading');
    expect(headerEls).toHaveLength(1);

    const headerEl = headerEls[0];
    expect(headerEl.classList).toContain('ds-c-accordion__heading');
  });

  it('renders accordion content', () => {
    renderAccordionItem();

    const contentEl = screen.getByText(defaultChildren);
    expect(contentEl.classList).toContain('ds-c-accordion__content');
  });

  it('renders additional className for header button', () => {
    renderAccordionItem({ buttonClassName: 'ds-u-test' });

    const buttonEl = screen.getByRole('button');
    expect(buttonEl.classList).toContain('ds-u-test');
  });

  it('renders additional className for content', () => {
    renderAccordionItem({ contentClassName: 'ds-u-test' });

    const contentEl = screen.getByText(defaultChildren);
    expect(contentEl.classList).toContain('ds-u-test');
  });

  it('renders header text', () => {
    renderAccordionItem({ heading: 'Foo' });

    const headingEl = screen.getByRole('heading');
    expect(headingEl.textContent).toBe('FooOpen');
  });

  it('renders an id automatically', () => {
    renderAccordionItem();

    const buttonEl = screen.getByRole('button');
    const contentEl = screen.getByText(defaultChildren);

    expect(buttonEl).toHaveAttribute('aria-controls');
    expect(buttonEl).toHaveAttribute('id');
    expect(contentEl).toHaveAttribute('aria-labelledby');
    expect(contentEl).toHaveAttribute('id');
  });

  it('renders a user set id ', () => {
    renderAccordionItem({ id: 'test-id' });

    const buttonEl = screen.getByRole('button');
    const contentEl = screen.getByText(defaultChildren);

    expect(buttonEl).toHaveAttribute('aria-controls', 'test-id');
    expect(buttonEl).toHaveAttribute('id', 'test-id-button');
    expect(contentEl).toHaveAttribute('aria-labelledby', 'test-id-button');
    expect(contentEl).toHaveAttribute('id', 'test-id');
  });

  it('renders an expanded or open accordion item', () => {
    renderAccordionItem({ defaultOpen: true });

    const buttonEl = screen.getByRole('button');
    expect(buttonEl).toHaveAttribute('aria-expanded', 'true');
  });

  it('renders an collapsed or closed accordion item', () => {
    renderAccordionItem({ defaultOpen: false });

    const buttonEl = screen.getByRole('button');
    expect(buttonEl).toHaveAttribute('aria-expanded', 'false');
  });
});

describe('Controlled accordion item', function () {
  it('renders button and should call onClick function when clicked', () => {
    const onClick = jest.fn();
    renderAccordionItem({ heading: 'Foo', onChange: onClick });

    const buttonEl = screen.getByRole('button');
    userEvent.click(buttonEl);
    expect(onClick).toHaveBeenCalled();
  });
});
