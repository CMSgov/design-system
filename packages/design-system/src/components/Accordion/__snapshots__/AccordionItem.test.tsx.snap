// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItem renders an accordion item 1`] = `
<DocumentFragment>
  <h2
    class="ds-c-accordion__heading"
  >
    <button
      aria-controls="static-id"
      aria-expanded="false"
      class="ds-c-accordion__button"
      id="static-id__button"
      type="button"
    >
      Heading for accordion item
      <svg
        aria-hidden="false"
        aria-labelledby="static-id__icon__title"
        class="ds-c-icon ds-c-icon--add ds-c-accordion__button-icon"
        id="static-id__icon"
        role="img"
        viewBox="3 3 18 18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title
          id="static-id__icon__title"
        >
          Open
        </title>
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </button>
  </h2>
  <div
    aria-labelledby="static-id__button"
    class="ds-c-accordion__content"
    hidden=""
    id="static-id"
  >
    &lt;p&gt;Content for accordion item&lt;/p&gt;
  </div>
</DocumentFragment>
`;
